File: src/main/java/com/example/javaee/entities/DatedModel.java
package com.example.javaee.entities;

import lombok.Data;
import lombok.EqualsAndHashCode;

import javax.persistence.Column;
import javax.persistence.MappedSuperclass;
import javax.persistence.PrePersist;
import javax.persistence.PreUpdate;
import java.util.Date;

@Data
@MappedSuperclass
@EqualsAndHashCode(callSuper = true)
public abstract class DatedModel extends Model {

	@Column(nullable = false)
	private Date created;

	@Column(nullable = false)
	private Date modified;

	@PrePersist
	public void create() {
		this.created = new Date();
	}

	@PreUpdate
	public void modify() {
		this.modified = new Date();
	}
}

File: src/main/kotlin/org/superbiz/javaee/entities/DatedModel.kt
package org.superbiz.javaee.entities

import javax.persistence.*
import java.util.*
import org.superbiz.javaee.entities.Model

MappedSuperclass
data public abstract class DatedModel : Model() {

    Column
    public var created: Date = Date()
    Column
    public var modified: Date = Date()

    PrePersist
    public fun create() {
        this.created = Date()
    }

    PreUpdate
    public fun modify() {
        this.modified = Date()
    }
}